Data Mapping Validation

alexandar1000 opened the pull request.

William-Blackie reviewed (changes requested):
Hi @alexandar1000, this looks good so far although I'm struggling to run this locally with this traceback;

`Traceback (most recent call last):
  File "/home/william/Projects/NSHCS-TMS/data-shaping/main.py", line 21, in <module>
    data_tables = read_spreadsheet(INPUT_DATA_TABLES_PATH)
  File "/home/william/Projects/NSHCS-TMS/data-shaping/ingest/ingest.py", line 16, in read_spreadsheet
    sheets[sheet_name] = pd.read_excel(xls, sheet_name=sheet_name,
  File "/home/william/Projects/NSHCS-TMS/data-shaping/.env/lib/python3.10/site-packages/pandas/io/excel/_base.py", line 486, in read_excel
    data = io.parse(
  File "/home/william/Projects/NSHCS-TMS/data-shaping/.env/lib/python3.10/site-packages/pandas/io/excel/_base.py", line 1551, in parse
    return self._reader.parse(
  File "/home/william/Projects/NSHCS-TMS/data-shaping/.env/lib/python3.10/site-packages/pandas/io/excel/_base.py", line 889, in parse
    raise err
  File "/home/william/Projects/NSHCS-TMS/data-shaping/.env/lib/python3.10/site-packages/pandas/io/excel/_base.py", line 876, in parse
    output[asheetname] = parser.read(nrows=nrows)
  File "/home/william/Projects/NSHCS-TMS/data-shaping/.env/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1704, in read
    ) = self._engine.read(  # type: ignore[attr-defined]
  File "/home/william/Projects/NSHCS-TMS/data-shaping/.env/lib/python3.10/site-packages/pandas/io/parsers/python_parser.py", line 293, in read
    conv_data = self._convert_data(data)
  File "/home/william/Projects/NSHCS-TMS/data-shaping/.env/lib/python3.10/site-packages/pandas/io/parsers/python_parser.py", line 364, in _convert_data
    return self._convert_to_ndarrays(
  File "/home/william/Projects/NSHCS-TMS/data-shaping/.env/lib/python3.10/site-packages/pandas/io/parsers/base_parser.py", line 551, in _convert_to_ndarrays
    values = lib.map_infer(values, conv_f)
  File "pandas/_libs/lib.pyx", line 2834, in pandas._libs.lib.map_infer
  File "/home/william/Projects/NSHCS-TMS/data-shaping/ingest/ingest.py", line 8, in clean_date
    return excel_date.date()
AttributeError: 'datetime.time' object has no attribute 'date' (sheet: tblTraineeContacts)`

Perhaps my test data is out of date or there's an env change that's not reflected in this repo?

alexandar1000 commented on .vscode/launch.json:
```diff
@@ -0,0 +1,19 @@
+{
```
I believed that it would be good to have a debugger set up from the get go. I know it is vscode specific but can still be useful. What do you think? From what I know it should work across devices.

alexandar1000 commented on data-shaping/data/new_model/.~lock.Database.xlsx#:
```diff
@@ -0,0 +1 @@
+,aleksandar,rubber-dinghy,28.12.2023 11:31,file:///home/aleksandar/.config/libreoffice/4;
```
We don't. It should not be in there. Deleted and gitignore omits them.

alexandar1000 commented on .vscode/settings.json:
```diff
@@ -0,0 +1,6 @@
+{
```
As opose to the previous one, this on can go. Deleted.

alexandar1000 commented on data-shaping/validator/schema.py:
```diff
@@ -0,0 +1,349 @@
+        "STID": pa.Column(int),
+        "STSTAT": pa.Column(str, checks=pa.Check.str_length(min_value=0, max_value=100)),
+        "STORDER": pa.Column(int)
+        # TODO: is_support is missing
```
I believe that we do not have from where to map this field. It is not present in the input data, but is in the output. We have a list of points that should be discussed in a meeting, one of which is this one.

alexandar1000 commented:
As the overall comment, the running does produce an error when validating as it has a few discrepancies among the model and data. Before this goes in, we would need to address those.

I will list them beneath to make it easier to prepare for the meeting:

On input:
:x: LeaveOfAbsenceRecord start_date -> tblSickLeave SLSTDT is not Nullable, but the data is missing/NULL
:x: SupportRecord open_date -> tblSupprt PSDT is not Nullable, but data is missing/NULL
:x: TraineeContact trainee_id -> tblTraineeContacts RGID is not Nullable, but data is missing/NULL
:x: Model shows ReviewProgression the sheets have AnnualReviewProgression and MidReviewProgression
:x: MidReviewProgression outcome_id -> tblMRP MROTCM is not Nullable, but data is missing/NULL
:x: EmploymentRecord start_date -> tblEmployers EMSTDT is not Nullable, but data is missing/NULL
:x: TrainingRecord job_sector -> tblRegistration RGFRJB is not Nullable, but data is missing/NULL
:x: Trainee status is extracted into a separate many to many table TraineeStatuses, but is kept in the Trainee table

On output:
:warning: Salary -> changed column from name to title to value with the model
:warning: JobSector -> changed column from name to name to title with the model
 :x: TraineeStatus -> is_support is missing
 :x: TraineeStatus -> enable_notifications is missing
 :x: SupportRecord -> support_status_id is missing
 :x: SupportRecord -> close_date is missing
 :x: SupportRecord -> open_date is not Nullable, but data is missing/NULL
 :x: TraineeContact -> trainee_id is not Nullable, but data is missing/NULL
 :x: LeaveOfAbsenceRecord -> start_date is not Nullable, but the data is missing/NULL
 :x: ReviewProgression -> not present in the dataframe, it is split between Annual and Mid ReviewProgressions
 :x: EmploymentRecord -> start_date  is not Nullable, but data is missing/NULL
 :x: EmploymentRecord -> finish_date  is not Nullable, but data is missing/NULL
 :x: EmploymentRecord -> employer_id  is not Nullable, but data is missing/NULL
 :x: TraineeStatuses -> support_record_id is missing

alexandar1000 commented:
After a meeting with the team, this is the agreed outcome to the abovementioned issues. Still to be implemented.

On input:
:x: LeaveOfAbsenceRecord start_date -> tblSickLeave SLSTDT is not Nullable, but the data is missing/NULL
- must allow missing data (populate with old data)
- is it going into the support?
:x: SupportRecord open_date -> tblSupprt PSDT is not Nullable, but data is missing/NULL
- sys default
:x: TraineeContact trainee_id -> tblTraineeContacts RGID is not Nullable, but data is missing/NULL
- delete these records (by Alex)
:x: Model shows ReviewProgression the sheets have AnnualReviewProgression and MidReviewProgression
- completely combine into the ReviewProgression
- will be merged in OutSystems
:x: MidReviewProgression outcome_id -> tblMRP MROTCM is not Nullable, but data is missing/NULL
- no outcome because trainee left the programme
- the rest are not certain
- TODO: Check if this is correct and how it should be handled
- TODO: Should not be nullable in OutSystems
:x: EmploymentRecord start_date -> tblEmployers EMSTDT is not Nullable, but data is missing/NULL
- Should be nullable as the start date is not necessarily know 
:x: TrainingRecord job_sector -> tblRegistration RGFRJB is not Nullable, but data is missing/NULL
- should be nullable as they are under no obligation to share this information
- TODO: PostTraining - check if there is duplicated
:x: Trainee status is extracted into a separate many to many table TraineeStatuses, but is kept in the Trainee table
- not on the trainee table any more
- TODO: Check this is consistent
On output:
:warning: Salary -> changed column from name to title to value with the model
- all good
:warning: JobSector -> changed column from name to name to title with the model
- all good
 :x: TraineeStatus -> is_support is missing
 - internal to outsystems
 :x: LeaveOfAbsenceRecord -> enable_notifications is missing
 - mark it as false initially
 - TODO: Figure out how to implement this tblMatPatLeave -> MLRTDT (figure out the time delta from this)
 :x: SupportRecord -> support_status_id is missing
 - internal to OutSystem
 :x: SupportRecord -> close_date is missing
 - internal to OutSystem
 :x: SupportRecord -> open_date is not Nullable, but data is missing/NULL
 - sys default
 :x: TraineeContact -> trainee_id is not Nullable, but data is missing/NULL
 - deleted by Alex
 :x: LeaveOfAbsenceRecord -> start_date is not Nullable, but the data is missing/NULL
 - should be nullable
 :x: ReviewProgression -> not present in the dataframe, it is split between Annual and Mid ReviewProgressions
 - internal to OutSystem
 :x: EmploymentRecord -> start_date  is not Nullable, but data is missing/NULL
 - should be nullable
 :x: EmploymentRecord -> finish_date  is not Nullable, but data is missing/NULL
 - should be nullable
 :x: EmploymentRecord -> employer_id  is not Nullable, but data is missing/NULL
 - should not be nullable because we cannot have a record without it
 - TODO: Alex to remedy by deleting the
 :x: TraineeStatuses -> support_record_id is missing
 - internal to OutSystem

alexandar1000 commented:
Made changes to answer the points raised above. The fixes to solutions are marked with a :wrench: 
On input:
:x: LeaveOfAbsenceRecord start_date -> tblSickLeave SLSTDT is not Nullable, but the data is missing/NULL
- must allow missing data (populate with old data)
- is it going into the support?
:wrench:: add the value on ingest in the converter

:x: SupportRecord open_date -> tblSupprt PSDT is not Nullable, but data is missing/NULL
- sys default
:wrench:: add the value on ingest in the converter

:x: TraineeContact trainee_id -> tblTraineeContacts RGID is not Nullable, but data is missing/NULL
- delete these records (by Alex)
:wrench:: Alex removed the bad rows

:x: Model shows ReviewProgression the sheets have AnnualReviewProgression and MidReviewProgression
- completely combine into the ReviewProgression
- will be merged in OutSystems
:wrench:: No action needed

:x: MidReviewProgression outcome_id -> tblMRP MROTCM is not Nullable, but data is missing/NULL
- no outcome because trainee left the programme
- the rest are not certain
- TODO: Should not be nullable in OutSystems
:wrench:: did not change the mermaid model as it has combined arp and mrp, made sure no null values are present

:x: EmploymentRecord start_date -> tblEmployers EMSTDT is not Nullable, but data is missing/NULL
- Should be nullable as the start date is not necessarily know
:wrench:: Made input validation nullable, changed the mermaid model

:x: TrainingRecord job_sector -> tblRegistration RGFRJB is not Nullable, but data is missing/NULL
- should be nullable as they are under no obligation to share this information
- check if there is duplicate data on PostTraining
:wrench:: Made input validation nullable, checked if duplicate on PostTraining
:question:: There is a record on PostTraining start_date, and on TrainingRecord start_month & start_year

:x: Trainee status is extracted into a separate many to many table TraineeStatuses, but is kept in the Trainee table
- not on the trainee table any more
- TODO: Check this is consistent
:wrench:: No change needed. For @Roman to assess if needed

On output:
:warning: Salary -> changed column from name to title to value with the model
- all good
:warning: JobSector -> changed column from name to name to title with the model
- all good
 :x: TraineeStatus -> is_support is missing
 - internal to outsystems
:wrench:: Removed from output validation. @Roman mentioned it is managed in OutSystems.

 :x: LeaveOfAbsenceRecord -> enable_notifications is missing
 - mark it as false initially
 - TODO: Figure out how to implement this tblMatPatLeave -> MLRTDT (figure out the time delta from this)
:wrench:: Removed from output validation. @Roman mentioned it is managed in OutSystems. No Idea what the TODO is.

 :x: SupportRecord -> support_status_id is missing
 - internal to OutSystem
:wrench:: Removed from output validation. @Roman mentioned it is managed in OutSystems.

 :x: SupportRecord -> close_date is missing
 - internal to OutSystem
:wrench:: Removed from output validation. @Roman mentioned it is managed in OutSystems.

 :x: SupportRecord -> open_date is not Nullable, but data is missing/NULL
 - sys default
:wrench:: Assigned the sysdefault value on data ingestion

 :x: TraineeContact -> trainee_id is not Nullable, but data is missing/NULL
 - deleted by Alex
:wrench:: Removed by Alex

 :x: LeaveOfAbsenceRecord -> start_date is not Nullable, but the data is missing/NULL
 - should be nullable
:wrench:: Made it nullable on output validation.

 :x: ReviewProgression -> not present in the dataframe, it is split between Annual and Mid ReviewProgressions
 - internal to OutSystem
:wrench:: No action needed. @Roman mentioned it is managed in OutSystems.

 :x: EmploymentRecord -> start_date  is not Nullable, but data is missing/NULL
 - should be nullable
:wrench:: Made it nullable on output validation.

 :x: EmploymentRecord -> finish_date  is not Nullable, but data is missing/NULL
 - should be nullable
:wrench:: Apparently already nullable. No action done.

 :x: EmploymentRecord -> employer_id  is not Nullable, but data is missing/NULL
 - should not be nullable because we cannot have a record without it
 - Alex to remedy by deleting the
:wrench:: Alex remedied it by deleting the affected rows.

 :x: TraineeStatuses -> support_record_id is missing
 - internal to OutSystem
:wrench:: Removed from output validation. @Roman mentioned it is managed in OutSystems.

alexandar1000 commented:
Moreover, the open questions are:

The issues for Alex:
1. CNENDT under tblTraineeContacts have years 2923 and 2922 under indexes 7008 and 7078
2. tblMRP MROTCM outcome still has empty values. Should it be null or values removed?
3. tblMRP and tblARP have empty dates and are nullable in the database. Is this ok or should it be non-nullable and casted to the 'null date'?
4. tblRegistration RGHEMTH and RGHCPATH should be integers but are dates in 1900s . How should we solve this? Currently, the function drops all values that are not integers
5. tblRegistration RGCERTDT has weird values such as the string '0224-02-19' . For now converted into a date 19/02/2024
6. tblRegistration RGMScDT has weird values such as the string '2921-11-17' . For now converted into a date 17/11/2017
7. tblRegistration RGMScOTCM again has dates in 1900s that should be an integer. Currently dropping the value if it is not an integer
8. tblRegistration RGNAT has a weird error that even though nullable, the values are registered as NaN. Currently dropping all values that are like this.
9. tblRegistration RGETHO has integers instead of descriptions in several places. What should be done with this? Not converting this as it should be fixed. I believe it should be written in the RGREL column.
10. tblRegistration RGHCSUP has some weird value dates in 1899s or 0. Needs to be cleaned. It causes garbage-in-garbage-out
